### DISCUSSION TITLE
Custom requests (via custom js function) which update http metrics

### DISCUSSION THREAD

**Discussion by Hamz4h** [artilleryio/artillery#1596] [Artillery]
Hey - I've got a requirement to parallel execute a number of the same requests (due to the design of the application under test) - the use case is a previous request will return an array of X and the number of elements in this array will dictate the number of parallel requests that will hit the backend API - this can be anywhere from 1 to 50 requests.

I couldn't get the `- parallel` approach to work so resorted to some custom js / `promise.all` to achieve what I wanted. This is working as expected and I can get a custom counter metric via the event emitter but what this leaves is the `http.codes.x` and `http.request_rate` in an inconsistent state. I would like to update the counter and request rate with the outcome of the parallel requests.

Whats the correct way to / correct events (or custom events) to account for this execution? I can directly update the `http.requests` counter but this doesn't update the remaining properties.

It's also not obvious how to update the `http.response_time` with this pool of requests. I may of course be approaching this the wrong way so I'm reaching out here.

Some code:

Note: I'm using `node-fetch` - realised that artillery uses `got` so will probably change this...

```
parallelRequests: (context, events, next) => {
        let endpoint = `http://some/end/point/here`;
        let requestCount = Math.floor(Math.random() * (50 - 30 + 1) + 30);

        (async () => {
            const concurrentReq = new Array(requestCount);

            for (let index = 0; index <= requestCount; index++) {
                concurrentReq.push(fetch(endpoint, {
                    method: 'POST',
                    body: JSON.stringify(somejsonbody())
                }).then(response => {
                    if (response.status != 200) {
                        throw new Error(`response_status_code_${response.status}`);
                    }
                }));
            }

            await Promise.all(concurrentReq);
        })().then(() => {
            events.emit("counter", "http.requests", requestCount); // update requests
            return next();
        }).catch(error => {
            return next(new Error(error));
        });
    }
```

**Comment by hassy** [comment]
hi @Hamz4h 👋 

> It's also not obvious how to update the http.response_time with this pool of requests

`http.response_time` is of type `histogram`, so you'd do something like this:

```js
events.emit('histogram', 'http.response_time', latencyInMs)
```

And those measurements would be recorded alongside the ones tracked automatically by Artillery.

https://www.artillery.io/docs/guides/guides/extension-apis#custom-metrics-api

**Reply by Hamz4h** [reply]
@hassy hey - finally had some time to give this a go - all looks good for the most part with the exception of `http.request_rate`. Adding to the request counter is fine, adding the response code counters is also fine and adding to the response_time histogram is working ok from what I can tell.

How would I get the `http.request_rate` to acknowledge the 4 concurrent requests that were sent for example? Here is a sample output I'm seeing

```
http.codes.200: ................................................................ 40
http.list_tables.requests: ..................................................... 30
http.request_rate: ............................................................. 1/sec
http.requests: ................................................................. 40
```

The request rate in this period is certainly not 1 a sec (at a min its 4 a second) - how do I go about making sure the request rate is accurate?

Edit: quick note - appreciate this is probably more a visual thing and I have plenty of stats outside of Artillery to refer to (which of course are all accurate as they come from the service under test) however, it would be nice to have consistent counters / rates across the different tooling in use.

Thanks for the help so far!

**Reply by hassy** [reply]
`http.request_rate` is tracked by Artillery's HTTP engine. As Artillery has no visibility into any requests that may be sent from your own JS code, you need to give it a hint. It's the same API as for histograms/counters, but with:

```js
events.emit('rate', 'http.request_rate');
```

If you make sure to emit a rate metric for each request your code is making, the numbers in the report should reflect the total number of requests sent.

One nuance with the rate metrics you might run into is that the number is averaged over the reporting period (of 10 seconds). So for example if 50 requests are sent in second 1, and nothing is sent in the next 9 seconds, the rate will be reported as 50/10 = 5. It's usually not an issue for most tests that run with continuous load, but something to be aware of.

**Reply by Hamz4h** [reply]
@hassy superstar thank you very much! Works great :)
